File: common-interpreters/shared/src/main/scala/co/topl/interpreters/InMemorySecureStore.scala
package co.topl.interpreters

import cats.Monad
import cats.data.Chain
import cats.effect.kernel.{Concurrent, Ref}
import cats.implicits._
import co.topl.algebras.SecureStore
import co.topl.codecs.bytes.typeclasses.Persistable
import co.topl.codecs.bytes.typeclasses.implicits._
import co.topl.models.Bytes

object InMemorySecureStore {

  object Eval {

    def make[F[_]: Monad: Concurrent]: F[SecureStore[F]] =
      Ref
        .of(Map.empty[String, Array[Byte]])
        .map(ref =>
          new SecureStore[F] {

            def write[A: Persistable](name: String, data: A): F[Unit] =
              for {
                _ <- erase(name)
                _ <- ref.update(m => m.updated(name, data.persistedBytes.toArray))
              } yield ()

            def consume[A: Persistable](name: String): F[Option[A]] =
              for {
                v <- ref.get.map(_.get(name).flatMap(Bytes(_).decodePersisted[A].toOption))
                _ <- erase(name)
              } yield v

            def list: F[Chain[String]] =
              ref.get.map(_.keys.toList).map(Chain.fromSeq)

            def erase(name: String): F[Unit] =
              ref.update(m =>
                m.updatedWith(name) { t =>
                  t.foreach { arr =>
                    arr.indices.foreach(arr(_) = 0)
                  }
                  None
                }
              )
          }
        )
  }
}

File: dart/minting/lib/interpreters/in_memory_secure_store.dart
import 'package:bifrost_minting/algebras/secure_store_algebra.dart';

class InMemorySecureStore extends SecureStoreAlgebra {
  final _store = <String, List<int>>{};
  @override
  Future<List<int>?> consume(String name) async {
    return _store[name];
  }

  @override
  Future<void> erase(String name) async {
    _store.remove(name);
  }

  @override
  Future<List<String>> list() async {
    return _store.keys.toList();
  }

  @override
  Future<void> write(String name, List<int> bytes) async {
    _store[name] = bytes;
  }
}
